uplink/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+const AWS_URL: &str = "https://aws.uplink.api.apollographql.com/graphql";
+
+#[derive(GraphQLQuery)]
+#[graphql(
+    query_path = "query.graphql",
+    schema_path = "uplink.graphql",
+    request_derives = "Debug",
+    response_derives = "PartialEq, Debug, Deserialize",
+    deprecated = "warn"
+)]
+
+pub(crate) struct SupergraphSdl;
+
+#[derive(Debug)]
+pub enum Error {
+    Reqwest(reqwest::Error),
+    EmptyResponse,
+    UpLink {
+        code: FetchErrorCode,
+        message: String,
+    },
+}
+
+impl From<reqwest::Error> for Error {
+    fn from(e: reqwest::Error) -> Self {
+        Error::Reqwest(e)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Schema {
+    pub id: String,
+    pub schema: String,
+}
+
+/// regularly download a schema from Uplink
+pub fn stream_supergraph(
+    api_key: String,
+    graph_ref: String,
+    interval: Duration,
+) -> impl Stream<Item = Result<Schema, Error>> {
+    let (sender, receiver) = channel(2);
+    let _ = tokio::task::spawn(async move {
+        let mut composition_id = None;
+
+        loop {
+            match fetch_supergraph(
+                api_key.to_string(),
+                graph_ref.to_string(),
+                composition_id.clone(),
+            )
+            .await
+            {
+                Ok(value) => match value.router_config {
+                    supergraph_sdl::SupergraphSdlRouterConfig::RouterConfigResult(
+                        schema_config,
+                    ) => {
+                        composition_id = Some(schema_config.id.clone());
+                        if sender
+                            .send(Ok(Schema {
+                                id: schema_config.id,
+                                schema: schema_config.supergraph_sdl,
+                            }))
+                            .await
+                            .is_err()
+                        {
+                            break;
+                        }
+                    }
+                    supergraph_sdl::SupergraphSdlRouterConfig::Unchanged => {
+                        tracing::trace!("schema did not change");
+                    }
+                    supergraph_sdl::SupergraphSdlRouterConfig::FetchError(e) => {
+                        if sender
+                            .send(Err(Error::UpLink {
+                                code: e.code,
+                                message: e.message,
+                            }))
+                            .await
+                            .is_err()
+                        {
+                            break;
+                        }
+                    }
+                },
+                Err(err) => {
+                    tracing::error!("error fetching supergraph from Uplink: {:?}", err);
+                    if sender.send(Err(err)).await.is_err() {
+                        break;
+                    }
+                }
+            }
+
+            tokio::time::sleep(interval).await
```

Comment:
Just a nit but I think https://docs.rs/tokio/latest/tokio/time/struct.Interval.html#method.tick could be more readable in that case instead of a loop with a sleep inside